src/main/java/net/snowflake/ingest/connection/OAuthManager.java

```diff
@@ -0,0 +1,177 @@
+   * @param telemetryService reference to the telemetry service
+   */
+  OAuthManager(
+      String accountName,
+      String username,
+      OAuthCredential oAuthCredential,
+      URIBuilder baseURIBuilder,
+      TelemetryService telemetryService) {
+    this(
+        accountName,
+        username,
+        oAuthCredential,
+        baseURIBuilder,
+        DEFAULT_UPDATE_THRESHOLD_RATIO,
+        telemetryService);
+  }
+
+  /**
+   * Creates a OAuthManager entity for a given account, user and OAuthCredential with a specified
+   * time to refresh the token
+   *
+   * @param accountName - the snowflake account name of this user
+   * @param username - the snowflake username of the current user
+   * @param oAuthCredential - the OAuth credential we're using to connect
+   * @param baseURIBuilder - the uri builder with common scheme, host and port
+   * @param updateThresholdRatio - the ratio between the expiration time of a token and the time
+   *     needed to refresh it.
+   * @param telemetryService reference to the telemetry service
+   */
+  OAuthManager(
+      String accountName,
+      String username,
+      OAuthCredential oAuthCredential,
+      URIBuilder baseURIBuilder,
+      double updateThresholdRatio,
+      TelemetryService telemetryService) {
+    // disable telemetry service until jdbc v3.13.34 is released
+    super(accountName, username, null);
+
+    // if any of our arguments are null, throw an exception
+    if (oAuthCredential == null || baseURIBuilder == null) {
+      throw new IllegalArgumentException();
+    }
+
+    // check if update threshold is within (0, 1)
+    if (updateThresholdRatio <= 0 || updateThresholdRatio >= 1) {
+      throw new IllegalArgumentException("updateThresholdRatio should fall in (0, 1)");
+    }
+    this.updateThresholdRatio = updateThresholdRatio;
+    this.oAuthClient = new SnowflakeOAuthClient(accountName, oAuthCredential, baseURIBuilder);
+
+    // generate our first token
+    refreshToken();
+  }
+
+  /**
+   * Creates a OAuthManager entity for a given account, user and OAuthClient with a specified time
+   * to refresh the token. Use for testing only.
+   *
+   * @param accountName - the snowflake account name of this user
+   * @param username - the snowflake username of the current user
+   * @param oAuthClient - the OAuth client to perform token refresh
+   * @param updateThresholdRatio - the ratio between the expiration time of a token and the time
+   *     needed to refresh it.
+   */
+  public OAuthManager(
+      String accountName, String username, OAuthClient oAuthClient, double updateThresholdRatio) {
+    super(accountName, username, null);
+
+    this.updateThresholdRatio = updateThresholdRatio;
+    this.oAuthClient = oAuthClient;
+
+    refreshToken();
+  }
+
+  @Override
+  String getToken() {
+    if (refreshFailed.get()) {
+      LOGGER.error("getToken request failed due to token refresh failure");
```

Comment:
which LOGGER are you using?

Reply:
And do we need to log it since an exception will be thrown?

Reply:
The LOGGER is from its super `SecurityManager`. Each client has exactly one SecurityManager which is either `JWTManager` or `OAuthManager`.